catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/CatalogServiceImpl.java

```diff
@@ -157,6 +160,79 @@ private IcebergStuff icebergStuff() {
         executor);
   }
 
+  @Override
+  public boolean validateStorageLocation(String location) {
+    StorageUri uri = StorageUri.of(location);
+    return objectIO.canResolve(uri);
+  }
+
+  @Override
+  public StorageUri locationForEntity(
+      WarehouseConfig warehouse,
+      ContentKey contentKey,
+      Content.Type contentType,
+      ApiContext apiContext,
+      String refName,
+      String hash) {
+    List<String> keyElements = contentKey.getElements();
+    int keyElementCount = keyElements.size();
+
+    List<ContentKey> keysInOrder = new ArrayList<>(keyElementCount);
+    for (int i = 0; i < keyElementCount; i++) {
+      ContentKey key = ContentKey.of(keyElements.subList(0, i + 1));
+      keysInOrder.add(key);
+    }
+
+    GetMultipleContentsResponse namespaces;
+    try {
+      namespaces =
+          contentService(apiContext)
+              .getMultipleContents(refName, hash, keysInOrder, false, API_READ);
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    Map<ContentKey, Content> contentsMap = namespaces.toContentsMap();
+
+    return locationForEntity(warehouse, contentKey, keysInOrder, contentsMap);
+  }
+
+  @Override
+  public StorageUri locationForEntity(
+      WarehouseConfig warehouse,
+      ContentKey contentKey,
+      List<ContentKey> keysInOrder,
+      Map<ContentKey, Content> contentsMap) {
+    List<String> keyElements = contentKey.getElements();
+    int keyElementCount = keyElements.size();
+    StorageUri location = null;
+    List<String> remainingElements = List.of();
+
+    // Find the nearest namespace with a 'location' property and start from there
+    for (int n = keysInOrder.size() - 1; n >= 0; n--) {
+      Content parent = contentsMap.get(keysInOrder.get(n));
+      if (parent != null && parent.getType().equals(NAMESPACE)) {
+        Namespace parentNamespace = (Namespace) parent;
+        String parentLocation = parentNamespace.getProperties().get("location");
+        if (parentLocation != null) {
+          location = StorageUri.of(parentLocation).withTrailingSeparator();
+          remainingElements = keyElements.subList(n + 1, keyElementCount);
+        }
+      }
+    }
+
+    // No parent namespace has a 'location' property, start from the warehouse
+    if (location == null) {
+      location = StorageUri.of(warehouse.location()).withTrailingSeparator();
+      remainingElements = keyElements;
+    }
+
+    for (String element : remainingElements) {
+      location = location.withTrailingSeparator().resolve(element);
```

Comment:
This is still risky as `element` might have `/` chars in it.

Is our intention to use a sub-dir per namespace? If so, should be error out in case `elements` contains a `/`?

Reply:
There are actually no "directories" in object stores - so `/` is rather just some character.

We'll need a follow-up to ensure unique / non-nested locations.

Reply:
The effect of `StorageUri.resolve()` depends on having `/` chars in `element`. In particular `StorageUri.of("s3://example/a/b/c/").resolve("/d")` will be `s3://example/d`... Are we ok with that for now?

Reply:
yep, should be good